docs(release): add v2.0.0 release notes (#11156)
spiceai/spiceai: docs/release_notes/v2.0.0.md

# Spice v2.0.0 (June 4, 2026)

> Spice v2.0.0 is the next major release of Spice and a major milestone in the project's development, advancing Spice from a single-node engine into a distributed, AI-native data and query platform. The release is headlined by **multi-node distributed query**, now generally available — multi-active, highly-available, and object-store-native, built on Apache Ballista — distributing both query execution and ingestion across executors with data-local routing and per-executor statistics for distributed join planning. Alongside it, the **Spice Cayenne data accelerator is generally available**, built on the Vortex compressed columnar format, with a high-throughput CDC write path, MERGE INTO, SQL-defined partitioning, inline writes, a dedicated compaction runtime, and write-path statistics for distributed join sizing. The engine also moves to **DataFusion v52** with sort pushdown, a rewritten merge join, and dynamic filters, and the **Spice CLI is rewritten in Rust** as a single self-contained binary.

Reviewer: 'AI-native' is over-used and meaningless IMO. Despite these being release notes, I feel like we should at least reference what types of workloads/use cases all of these features are in service of  i.e. enterprise, data-intensive agents

Reviewer: Would perhaps even add another sentence on why agents require such infrastructure.

> Spice v2.0.0 is the next major release of Spice and a major milestone in the project's development, advancing Spice from a single-node engine into a distributed, AI-native data and query platform. The release is headlined by **multi-node distributed query**, now generally available — multi-active, highly-available, and object-store-native, built on Apache Ballista — distributing both query execution and ingestion across executors with data-local routing and per-executor statistics for distributed join planning. Alongside it, the **Spice Cayenne data accelerator is generally available**, built on the Vortex compressed columnar format, with a high-throughput CDC write path, MERGE INTO, SQL-defined partitioning, inline writes, a dedicated compaction runtime, and write-path statistics for distributed join sizing. The engine also moves to **DataFusion v52** with sort pushdown, a rewritten merge join, and dynamic filters, and the **Spice CLI is rewritten in Rust** as a single self-contained binary.

Reviewer: ```suggestion
Spice v2.0.0 is the next major release of Spice and a major milestone in the project's development, advancing Spice from a single-node engine into a distributed data and query platform built for enterprise AI agents. The release is headlined by **multi-node distributed query**, now generally available — multi-active, highly-available, and object-store-native, built on Apache Ballista — distributing both query execution and ingestion across executors with data-local routing and per-executor statistics for distributed join planning. Alongside it, the **Spice Cayenne data accelerator is generally available**, built on the Vortex compressed columnar format, with a high-throughput CDC write path, MERGE INTO, SQL-defined partitioning, inline writes, a dedicated compaction runtime, and write-path statistics for distributed join sizing. The engine also moves to **DataFusion v52** with sort pushdown, a rewritten merge join, and dynamic filters, and the **Spice CLI is rewritten in Rust** as a single self-contained binary.
```

> v2.0 also expands real-time and write-path capabilities across the platform: native **CDC from MongoDB Change Streams and PostgreSQL WAL logical replication**, durable **Kafka CDC offsets**, **DML write-back** for PostgreSQL, Snowflake, DynamoDB, Arrow, and DuckLake, **DDL and MERGE INTO** for Iceberg and Cayenne catalogs, **mutual TLS** across server endpoints and outbound connectors, **HashiCorp Vault and Azure Key Vault secret stores**, **user-defined functions**, hybrid **search with Elasticsearch and DuckDB HNSW vector indexes**, provider-aware **LLM prompt caching**, and the **Responses API across all model providers**.

Author: ```suggestion
v2.0 also expands real-time and write-path capabilities across the platform: native **CDC from MongoDB Change Streams and PostgreSQL WAL logical replication**, durable **Kafka CDC offsets**, **DML write-back** for PostgreSQL, Snowflake, DynamoDB, Arrow, and DuckLake, **DDL and MERGE INTO** for Iceberg catalogs, **mutual TLS** across server endpoints and outbound connectors, **HashiCorp Vault and Azure Key Vault secret stores**, **user-defined functions**, hybrid **search with Elasticsearch and DuckDB HNSW vector indexes**, provider-aware **LLM prompt caching**, and the **Responses API across all model providers**.
```

- **[Multi-Active HA Distributed Query (GA)](#multi-active-ha-distributed-query-ga)** — multi-node distributed query built on Apache Ballista, with object-store-native clustering, dynamic cluster sizing, distributed ingestion, data-local query routing, per-executor table statistics for distributed join planning, and async queries via `/v1/queries`
- **[Mutual TLS (mTLS)](#security-mutual-tls-secret-stores-and-hardening)** — public mTLS for HTTP and Flight, TLS cert hot-reload, and mTLS client certificates for FlightSQL and Spice.ai connectors
- **[Enterprise Authentication & Authorization](#security-mutual-tls-secret-stores-and-hardening)** — OIDC bearer-token verification and Cedar-based authorization policy with per-principal row- and column-level filtering
- **[New Secret Stores](#security-mutual-tls-secret-stores-and-hardening)** — HashiCorp Vault and Azure Key Vault
- **[CDC Sources](#change-data-capture-cdc-sources)** — native MongoDB Change Streams, PostgreSQL WAL logical replication, and durable Kafka CDC offsets — no Debezium or Kafka middleware required

> - **[DML & DDL](#dml-ddl-and-write-back)** — INSERT/UPDATE/DELETE write-back for PostgreSQL, Snowflake, DynamoDB, and Arrow; `CREATE TABLE`/`DROP TABLE` and `MERGE INTO` for Iceberg and Cayenne catalogs

Author: ```suggestion
- **[DML & DDL](#dml-ddl-and-write-back)** — INSERT/UPDATE/DELETE write-back for PostgreSQL, Snowflake, DynamoDB, and Arrow; `CREATE TABLE`/`DROP TABLE` and `MERGE INTO` for Iceberg catalogs
```

- **Inline Writes**: Small writes are serialized as Arrow IPC and committed directly into the Cayenne metastore, bypassing the staged Vortex write path for low-latency ingest. Inline upserts atomically rewrite existing inline rows, inline data stays query-visible via an in-memory union scan, and rows are checkpointed to Vortex when thresholds are reached. Inline writes now also proceed with pending deletions in flight, and inline flush caps scale with available memory and storage class.
- **Fast-Path CDC Deletes**: `DELETE` statements whose filters identify primary keys directly — including composite keys expressed as `(k1, k2) IN ((...), (...))` — skip the table scan entirely.
- **Merge-On-Read Position Deletes**: Primary-key upsert tables use position deletes with memory-pool accounting, avoiding full-table rewrites on update-heavy workloads.
- **Resident Upsert Keysets**: CDC upsert primary-key keysets stay resident between batches, avoiding per-batch full-table rebuilds.
- **CDC Sub-Batch Efficiency**: Interleaved upsert/delete workloads produce fewer sub-batch splits, with last-write-wins deduplication applied within batches.

> - **Dedicated Compaction Runtime**: Background compaction runs on a dedicated runtime with CDC pipelining and protected snapshots, isolating compaction work from query and ingest paths.

Author: ```suggestion
- **Dedicated Compaction Runtime**: Background compaction runs on a dedicated thread pool with CDC pipelining and protected snapshots, isolating compaction work from query and ingest paths.
```

- **Scan-Result Cache**: A new scan-result cache accelerates hot reads, with parallel Vortex partition writes and lock-free deletion caches with bloom-prefiltered probes.

**SQL & catalog**:

- **MERGE INTO**: Upsert-style `MERGE INTO` for Cayenne catalog tables, distributed across executors in cluster mode.
- **`PARTITION BY` in SQL**: Define partitioning directly in `CREATE TABLE ... PARTITION BY (...)`; metadata is persisted in the catalog and survives restarts.
- **Composite Partitioning**: `partition_by: [col1, col2]` with hierarchical path-like keys.
- **File-Based Retention Deletes**: Time-based retention uses file-level deletes for both position-based and primary-key tables.

**Correctness**: Synchronized partition commits, correct `NULL`-sentinel handling for nullable partition expressions, tombstoned inline-checkpointed rows on upsert (preventing duplicate primary keys), and live reads through expired protected snapshots.

### Multi-Active HA Distributed Query (GA)

> [Spice.ai Enterprise](https://docs.spice.ai/docs/enterprise) feature. See [High Availability](https://docs.spice.ai/docs/enterprise/production/high-availability).

> [Distributed Query](https://spiceai.org/docs/features/distributed-query) is generally available with multi-active highly-available, object-store-native distributed query.

Reviewer: A bit redundant (starts and ends with distributed query), and just lists off attributes repeated from the title without explanation

> [Distributed Query](https://spiceai.org/docs/features/distributed-query) is generally available with multi-active highly-available, object-store-native distributed query.

Reviewer: ```suggestion
[Distributed Query](https://spiceai.org/docs/features/distributed-query) is generally available. Built on Apache Ballista, it distributes query execution across multiple active executor nodes with no single point of failure, reading directly from object storage rather than relying on a central cluster.
```

```yaml
datasets:
  - from: postgres:my_table
    name: my_table
    params:
      pg_host: localhost
      pg_db: mydb
    acceleration:
      enabled: true
      engine: duckdb
      refresh_mode: changes
```

### DML, DDL, and Write-Back

Spice v2.0 turns more connectors and catalogs into full read/write tables:

- **PostgreSQL DML**: `INSERT`, `UPDATE`, and `DELETE` write-back on [PostgreSQL](https://spiceai.org/docs/components/data-connectors/postgres) datasets, with foreign-key metadata exposed via the PostgreSQL catalog connector.
- **Snowflake DML**: `INSERT`, `UPDATE`, and `DELETE` write-back on [Snowflake](https://spiceai.org/docs/components/data-connectors/snowflake) datasets.
- **DynamoDB DML**: `INSERT`, `UPDATE`, and `DELETE` for [DynamoDB](https://spiceai.org/docs/components/data-connectors/dynamodb), complementing read and CDC streaming.
- **Arrow Primary Key Upserts**: Native update-or-insert semantics for in-memory [Arrow](https://spiceai.org/docs/components/data-accelerators/arrow)-accelerated tables.

> - **DDL for Iceberg and Cayenne**: `CREATE TABLE` and `DROP TABLE` via FlightSQL and `/v1/sql` for [Iceberg](https://iceberg.apache.org/) and Cayenne catalogs, with `catalog.access: read_write_create`.

Author: ```suggestion
- **DDL for Iceberg**: `CREATE TABLE` and `DROP TABLE` via FlightSQL and `/v1/sql` for [Iceberg](https://iceberg.apache.org/), with `catalog.access: read_write_create`.
```

### Dependency Updates

| Dependency / Component | Version                             |
| ---------------------- | ----------------------------------- |
| **DataFusion**         | v52.5                               |
| **Ballista**           | v52                                 |
| **Arrow (arrow-rs)**   | v57.2                               |
| **DuckDB**             | v1.5.3 (with statically linked VSS) |
| **iceberg-rust**       | v0.9.1                              |
| **Turso (libsql)**     | v0.6.1                              |
| **Vortex**             | v0.69.0                             |
| **delta_kernel**       | v0.18.2                             |
| **rmcp (MCP)**         | v1.5.0                              |
| **mistral.rs**         | v0.8.x (candle v0.10.1)             |
| **ADBC Core**          | v0.23                               |
| **Rust toolchain**     | v1.94.1                             |

## Contributors

> - [@avishafir](https://github.com/avishafir)

Author: ```suggestion
```

- [@cluster2600](https://github.com/cluster2600)
- [@ewgenius](https://github.com/ewgenius)
- [@Jeadie](https://github.com/Jeadie)
- [@krinart](https://github.com/krinart)
- [@lukekim](https://github.com/lukekim)
- [@peasee](https://github.com/peasee)
- [@penberg](https://github.com/penberg)
- [@phillipleblanc](https://github.com/phillipleblanc)
- [@sgrebnov](https://github.com/sgrebnov)

> - [@vyershov](https://github.com/vyershov)

Author: ```suggestion
```

## Breaking Changes

- **Models included by default**: The separate `models` build variant has been removed. Local LLM inference is always included in the default build and image.
- **Windows native builds removed**: Use [WSL](https://learn.microsoft.com/en-us/windows/wsl/) for local development.
- **Spicepod version defaults to `v2`**: `spice init` creates `version: v2` spicepods. `v1` remains supported with auto-migration; `v1beta1` is no longer accepted.

> - **Rust crate feature defaults**: Default runtime crate features were removed — `spiced` now enables features explicitly. Building from source with `--no-default-features` semantics may require updating feature flags.

Author: ```suggestion
```

- **Flattened `runtime.scheduler` configuration**: The nested `runtime.scheduler.partition_management` block is flattened and renamed:

  ```yaml
  # Before
  runtime:
    scheduler:
      partition_management:
        interval: 30s
        max_assignments_per_cycle: 16
        discovery_timeout: 10s

  # After
  runtime:
    scheduler:
      partition_assignment_interval: 30s
      max_assignments_per_interval: 16
      partition_discovery_timeout: 10s
  ```

- **S3 metadata columns renamed**: `location`, `last_modified`, `size` → `_location`, `_last_modified`, `_size`.
- **Default query memory limit changed**: Increased from 70% to 90%.
- **Metric renames**: `accelerated_refresh` metrics renamed to `acceleration_refresh`; `last_refresh_time` gauge renamed to include the milliseconds unit.

> - **Caching config renamed**: `ResultsCache` replaced with `SQLResultsCacheConfig`.

Author: ```suggestion
```

- **`/v1/search` API**: Always returns an array in `matches`, even for single results.
- **`/v1/evals` API removed**.
- **Perplexity model provider removed**.
- **x.ai model endpoint**: x.ai models exclusively use the `/v1/responses` endpoint.

## Upgrade Guide from v1.x

Most v1 spicepods continue to work on v2.0 — `v1` remains supported and deprecated fields auto-migrate at load time — so many deployments can upgrade by updating the binary or image alone. The steps below cover the [breaking changes](#breaking-changes) that may require manual action. Review each before upgrading a production deployment.

### 1. Build, image, and platform changes

- **Models are now included by default.** The separate `models` build variant (and the corresponding `-models` image tags) has been removed; local LLM inference is always included in the default build and image. If your deployment pinned a `models` build or `-models`-tagged image, switch to the default build/image.
- **Native Windows builds are removed.** Use [WSL](https://learn.microsoft.com/en-us/windows/wsl/) for local Windows development.

> - **Building from source:** Default runtime crate features were removed — `spiced` now enables features explicitly. If you build Spice from source, set the feature flags you need (see the `Makefile` build targets) rather than relying on default crate features.

Author: ```suggestion
```

| v1 (deprecated)               | v2 (preferred)                                                |
| ----------------------------- | ------------------------------------------------------------- |
| `runtime.results_cache`       | `runtime.caching.sql_results` (`cache_max_size` → `max_size`) |
| `runtime.memory_limit`        | `runtime.query.memory_limit`                                  |
| `runtime.temp_directory`      | `runtime.query.temp_directory`                                |
| `dataset.invalid_type_action` | `dataset.unsupported_type_action`                             |

### 3. Update changed configuration

- **Flattened `runtime.scheduler` configuration** (distributed deployments): the nested `partition_management` block is flattened and renamed. This is **not** auto-migrated — update it manually:

  ```yaml
  # Before (v1.x)
  runtime:
    scheduler:
      partition_management:
        interval: 30s
        max_assignments_per_cycle: 16
        discovery_timeout: 10s

  # After (v2.0)
  runtime:
    scheduler:
      partition_assignment_interval: 30s
      max_assignments_per_interval: 16
      partition_discovery_timeout: 10s
  ```

> 

Author: ```suggestion
```

- Implement DML for DynamoDBTableProvider by [@lukekim](https://github.com/lukekim) in [#9915](https://github.com/spiceai/spiceai/pull/9915)
- chore: Update iceberg-rust fork to v0.9 by [@lukekim](https://github.com/lukekim) in [#9917](https://github.com/spiceai/spiceai/pull/9917)
- Run physical optimizer on `FallbackOnZeroResultsScanExec` fallback plan by [@sgrebnov](https://github.com/sgrebnov) in [#9927](https://github.com/spiceai/spiceai/pull/9927)
- Improve Databricks error message when dataset has no columns by [@sgrebnov](https://github.com/sgrebnov) in [#9928](https://github.com/spiceai/spiceai/pull/9928)
- Delta Lake: fix data skipping for >= timestamp predicates by [@sgrebnov](https://github.com/sgrebnov) in [#9932](https://github.com/spiceai/spiceai/pull/9932)
- fix: Ensure distributed Cayenne DML inserts are forwarded to executors by [@Jeadie](https://github.com/Jeadie) in [#9948](https://github.com/spiceai/spiceai/pull/9948)
- Add full query federation support for ADBC data connector by [@lukekim](https://github.com/lukekim) in [#9953](https://github.com/spiceai/spiceai/pull/9953)

> - Make time_format deserialization case-insensitive by [@vyershov](https://github.com/vyershov) in [#9955](https://github.com/spiceai/spiceai/pull/9955)

Author: ```suggestion
- Make time_format deserialization case-insensitive by [@claudespice](https://github.com/claudespice) in [#9955](https://github.com/spiceai/spiceai/pull/9955)
```
